enhancements/mirror-registry-upgrade.md

```diff
@@ -0,0 +1,183 @@
+---
+title: Mirror Registry Upgrade
+authors:
+  - "@HammerMeetNail"
+reviewers:
+  - "@joking"
+approvers:
+  - "@joking"
+creation-date: 2022-01-20
+last-updated: 2022-01-20
+status: provisional
+---
+
+# mirror-registry-upgrade
+The mirror registry for Red Hat OpenShift is meant to be a long live registry that lives on a single RHEL host. It supports mirror for the last three version of OCP and must be able to handle upgrades gracefully. This proposal details how mirror registry will be able to upgrade to new versions, e.g., 1.0 -> 1.1.0. 
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+ > 1. When should an upgrade occur?
+ > 2. How should we handle certificate renewal?
+ > 3. Should we add a new command for upgrade of should install command trigger upgrade?
+ > 4. What would a self upgrading mirror registry look like in a disconnected environment?
+
+## Summary
+The mirror registry will need to be upgraded whenever a new y-stream release of `Quay` or `OCP` occurs. It should be upgraded whenever a z-stream release occurs for `Quay`, `Postgres`, or `Redis`. Certificates will be renewed at time of upgrade, which should occur at least once per year. Existing installation will be detected and no additional information should be required for upgrade. Upgrades will be idempotent. 
+
+### Goals
+
+* Users should be able to upgrade to newer versions of mirror registry with ease
+* Existing mirror registry installations should be detected and settings inherited
+* Autogenerated certificates should be renewed on upgrade
+
+### Non-Goals
```

Comment:
I'm not sure if you want to consider it in-scope for this particular proposal or not, but another maintenance task folks will want to perform with a long-running registry is pruning old images to avoid unbounded storage growth.  In the OpenShift-mirror context, this looks like:

1. User decides to install OpenShift release oA.
2. User installs a new mirror registry, which happens to be version mA.
3. New OpenShift releases oB, oC, ... come out and get mirrored into the mirror registry.
4. New mirror-registry releases mB, mC, ... come out, and the admin updates their mirror to use those versions.
5. User realizes that they really don't need oA and oB in the mirror registry anymore, so they unpin them.
6. Registry reaper removes all the layers and manifests and whatnot that were only needed for those unpinned OpenShift releases.

(4) is definitely in-scope for this enhancement.  I'm wondering whether (5) is in-scope.  I assume (6) would work out of the box, but if not, I'm curious about whether that's in-scope here too.

(5) isn't strictly unique to the mirror-registry workflow.  Folks might theoretically want that with highly-available registry implementations as well.  But while a few GiB of stale content on a highly-available registry isn't that big a deal, that same amount of cruft on the single-node mirror-registry deployment seems like it might be something that folks would care about.  And that's especially likely once we're throwing around [TiBs][1].

[1]: https://github.com/openshift/openshift-docs/pull/39593/files#diff-685e844b2628298265fd62c6ab6aa264d3f356a0ec70906362e13267c4f9f806R85